src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/MapCode/MapCodeEndpoint.cs

```diff
@@ -0,0 +1,557 @@
+                                Uri = documentContext.Uri
+                            },
+                            Edits = [edit],
+                        };
+
+                        changes.Add(textDocumentEdit);
+                        mappingSuccess = true;
+                    }
+                }
+
+                // We were able to successfully map using this focusLocation.
+                if (mappingSuccess)
+                {
+                    return;
+                }
+            }
+        }
+    }
+
+    private static async Task<(RazorProjectEngine projectEngine, List<RazorSourceDocument> importSources)> InitializeProjectEngineAsync(IDocumentSnapshot originalSnapshot)
+    {
+        var engine = originalSnapshot.Project.GetProjectEngine();
+        var importSources = new List<RazorSourceDocument>();
+
+        var imports = originalSnapshot.GetImports();
+        foreach (var import in imports)
+        {
+            var sourceText = await import.GetTextAsync().ConfigureAwait(false);
+            var source = sourceText.GetRazorSourceDocument(import.FilePath, import.TargetPath);
+            importSources.Add(source);
+        }
+
+        return (engine, importSources);
+    }
+
+    private static List<SyntaxNode> ExtractValidNodesToMap(SyntaxNode rootNode)
+    {
+        var validNodesToMap = new List<SyntaxNode>();
+        var stack = new Stack<SyntaxNode>();
+        stack.Push(rootNode);
+
+        while (stack.Count > 0)
+        {
+            var currentNode = stack.Pop();
+
+            if (s_validNodesToMap.Contains(currentNode.GetType()))
+            {
+                validNodesToMap.Add(currentNode);
+                continue;
+            }
+
+            // Add child nodes to the stack in reverse order for depth-first search
+            foreach (var childNode in currentNode.ChildNodes().Reverse())
+            {
+                stack.Push(childNode);
+            }
+        }
+
+        return validNodesToMap;
+    }
+
+    // These are the nodes that we currently support for mapping. We should update
+    // this list as the client evolves to send more types of nodes.
+    private readonly static List<Type> s_validNodesToMap =
+    [
+        typeof(CSharpCodeBlockSyntax),
+        typeof(CSharpExplicitExpressionSyntax),
+        typeof(CSharpImplicitExpressionSyntax),
+        typeof(MarkupElementSyntax),
+        typeof(MarkupTextLiteralSyntax),
+        typeof(RazorDirectiveSyntax),
+    ];
+
+    private async Task<bool> TrySendCSharpDelegatedMappingRequestAsync(
+        TextDocumentIdentifierAndVersion textDocumentIdentifier,
+        SyntaxNode nodeToMap,
+        LSP.Location[][] focusLocations,
+        ImmutableArray<TextDocumentEdit>.Builder changes,
+        CancellationToken cancellationToken)
+    {
+        var delegatedRequest = new DelegatedMapCodeParams(
+            textDocumentIdentifier,
+            RazorLanguageKind.CSharp,
+            [nodeToMap.ToFullString()],
```

Comment:
This is essentially an internal call, so if we're never going to send multiple things, it doesn't need to be an array